### DISCUSSION TITLE
[9일차 -윤경]es5와 es6의 차이점

### DISCUSSION THREAD

**Discussion by skyhanull** [zzangCS/zzangInterview#31] [프론트 면접 질문]
es5와 es6의 차이점은 무엇이 있을까요?

**Comment by skyhanull** [comment]
es5와 es6의 차이점이라고 한다면

let,const의 등장 => var가 아닌 let과 const의 등장으로 var의 재할당/재선언성을 보완하고자 탄생
Arrow Functions의 사용 => Arrow Functions은 자신의 this, arguments을 바인딩하지 않습니다
기존의 반복문에서 for..in/for..of문과 forEach문이 등장했습니다
구조분해할당과 spread연산자가 등장했습니다
템플릿 리터럴의 등장
비교문이 기존에는 '=='였지만 es6로 오면서 '==='로 바뀌게 되어 값만 아니고 자료형도 비교하게 되었습니다
promise , async/await의 등장 특히 async/await의 등장으로 비동기를 동기문처럼 쓸수 있게 되었습니다

**Comment by ez1n** [comment]
ES : ECMAScript의 약자 - 자바스크립트의 표준, 규격을 의미한다.
⇒ es5는 2009년, es6는 2015년에 출시되었다.

**차이점**

1. 변수 선언 - let, const 키워드
 블록 레벨 스코프를 가지는 let,과 const로 var 키워드의 함수 레벨 스코프와 암묵적 재할당으로 인한 단점들을 보완하였다.
    
2. arrow function(화살표 함수)
 화살표 함수를 통해 함수를 간결하게 나타낼 수 있어 가독성 및 유지보수성이 올라갔으나 기존 함수와 this 바인딩이 다르다. (this 존재 X → 상위 환경에서의 this 참조)
    
3. default parameter(매개변수 초깃값)
 기존 es5에서는 매개변수에 초깃값을 작성하려면 함수 내부에서 설정해주는 로직이 필요했으나 기본 초깃값이 추가되면서 그럴 필요가 없어졌다.
    
4. template literal (``) / multi-line string (줄바꿈)
 template literal이 도입되면서 문자열 안에서 자바스크립트 코드를 사용하는 것이 가능해 졌고 공백이나 줄바꿈을 사용하는 불편함이 사라졌다.
    
5. 클래스
 기존 자바스크립트에서는 프로토타입을 이용하여 class처럼 사용하였는데 es6에서 클래스 키워드가 도입되면서 이를 통해 선언할 수 있게 되었다.
    
6. 구조분해할당
 객체나 배열에서 자료구조를 파괴하여 개별 변수에 할당하는 것
⇒ 배열에서는 순서가 중요하며 객체에서는 키 값이 중요하다!
    
7. Promise
es6 이전 버전에서는 비동기 통신에 있어 콜백 함수를 사용하였는데 프로미스가 도입되면서 async, await을 통해 비동기 통신을 동기적으로 처리할 수 있게 되었다.
